Preparatory clean-up of ARC nSIM board before adding ARC HS cores support

abrodkin opened the pull request:
What we do in this series:
1. Clean-up of copy-pastes from EMSK board
2. Rename of `nsim_em` board to more generic `nsim` so it becomes usable for ARC HS cores
3. Move CPU selection to defconfigs removing implicit selection of default something

abrodkin committed:
soc/nsim: Minor clean-up

1. Rename "nsim" to the canonical nSIM, see [1]
2. Fix SoC description - it's obviously not EMSK

[1] https://www.synopsys.com/dw/ipdir.php?ds=sim_nSIM

Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

abrodkin committed:
simulation: ARC EM in nSIM to use nsim_em.props

This matches a naming scheme used for other flavour of ARC EM
SoC flavors.

Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

abrodkin committed:
board/nsim: Make it CPU-family independent

ARC nSIM simulates all flavors or ARC cores so there's
no point in limiting its usage to ARC EM family only.

Moreover with upcoming addition of ARC HS family support
in Zephyr we'll be re-using nSIM "board" for them as well.

Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

abrodkin committed:
board/nsim: Move CPU selection in defconfigs

This will allow us to easily specify other CPUs looking
forward and not rely on any default value.

Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

mbolivar reviewed (approved):
LGTM

dbkinder commented on boards/arc/nsim/Kconfig.board:
```diff
@@ -0,0 +1,13 @@
+	bool "ARC nSIM simulator"
+	depends on SOC_NSIM
+	help
+	  The DesignWare ARC nSIM board is a virtual board based on
```
How about:
```
          The DesignWare ARC nSIM board is a virtual board based on
	  the ARC nSIM simulator.  It demonstrates the ARC core features
	  and a console based on the hostlink feature of the nSIM simulator.
```

abrodkin replied:
Well nSIM simulates among other things legacy ARC UART which is used for serial console in Zephyr/Linux/U-Boot and this is a completely different beast as compared to so-called hostlink (which is just a mailbox in target's memory).

Makes sense?

dbkinder replied:
I was trying to clarify the English a bit.  You wrote, "It just has a console based on hostlink feature of nSIM simulator."  Is what I suggested incorrect?  Can you use the simulated UART for console output?

abrodkin replied:
@dbkinder Oops, sorry. I blindly fixed spelling of `nSIM` in original description (https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arc/nsim_em/Kconfig.board) and failed to realize description by its nature was misleading. Indeed we use simulated UART (ARC UART, see https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/serial/Kconfig.nsim).

So based on your suggestion I'll reword the comment like that:
> The DesignWare ARC nSIM board is a virtual board based on
> the ARC nSIM simulator.  It demonstrates the ARC core features
> and a console based on the legacy ARC UART model.

abrodkin commented:
@dbkinder pushed updated version. And thanks for the feedback!

abrodkin committed:
ARC: Mention ARC EM family instead of EM4 template

ARC EM4 is just a baseline configuration of ARC EM family of CPU cores.
But with addition of more featuers like caches, DSP extensions etc
we're effectively getting EM6, EM5D etc templates.

So to not confuse users let's talk about families of ARC cores
as that's what makes sense together with extra features but not
templates itself.

Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

vonhust reviewed (approved):
I am ok. It's a necessary cleanup before we add ARC HS support

ruuddw reviewed (approved):
I expect a lot of commonality between the nSIM EM and HS targets, makes sense to generalize instead of duplicate.

abrodkin committed:
board/nsim: Refactor device tree description

Instead of blind copy of nsim_em.dts re-structure this way:
 * nsim.dtsi     - Top-level "board" description re-usable for
 |                 all nSIM-based "boards".
 |
 |                 Even though it's not needed right now but it
 |                 allows to add other ARC core families in the future.
 |
 \_ nsim_em.dtsi - Common definitions for boards with ARC EM cores
    |
    \_ Real boards with ARC EM cores
      |
      \ nsim_em.dts
      \ nsim_sem.dts
      \ nsim_em_mpu_stack_guard.dts
      \ nsim_sem_mpu_stack_guard.dts

Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

abrodkin commented:
Yet another respin with rework of device tree description for:
1. Board-specific compatible & model
2. Ability to reuse .dtsi for other nSIM board flavors (read to be added ARC HS variants)

dbkinder reviewed (approved):
LGTM, thanks!

nashif commented:
issues in ci, for example:

```
In file included from ../../../../../../include/kernel_includes.h:21,
                 from ../../../../../../include/kernel.h:17,
                 from ../../../../../../kernel/include/kernel_structs.h:10,
                 from /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/arch/arc/core/offsets/offsets.c:26:
../../../../../../include/atomic.h:94:28: note: previous definition of 'atomic_sub' was here
 static inline atomic_val_t atomic_sub(atomic_t *target, atomic_val_t value)
```

abrodkin commented:
@nashif  

> issues in ci, for example:
> 
> ```
> In file included from ../../../../../../include/kernel_includes.h:21,
>                  from ../../../../../../include/kernel.h:17,
>                  from ../../../../../../kernel/include/kernel_structs.h:10,
>                  from /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/arch/arc/core/offsets/offsets.c:26:
> ../../../../../../include/atomic.h:94:28: note: previous definition of 'atomic_sub' was here
>  static inline atomic_val_t atomic_sub(atomic_t *target, atomic_val_t value)
> ```

Silly me. That was due to the following hunk that was supposed to get in later (and in a different form when adding ARC HS support):
```diff
--- a/arch/arc/Kconfig
+++ b/arch/arc/Kconfig

+config CPU_ARCHS
+       bool
+       default y
+       select CPU_ARCV2
+       select ATOMIC_OPERATIONS_BUILTIN
+       help
+         This option signifies the use of an ARC HS CPU
 endmenu
```

Given it was a copy-paste of `CPU_ARCEM` it has 2 things which when combined together give us the problem:
1. "default y" selecting this option as it was situated below `CPU_ARCEM`
2. select `ATOMIC_OPERATIONS_BUILTIN` which selects GCC's built-in atomic ops which are missing when building with `-mcpu=em4_fpus` as ARC EM family doesn't support `llock`/`scond` atomics.

Will push a fixed respin.
